dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/DefaultTrackedEntityService.java

```diff
@@ -253,35 +233,117 @@ public TrackedEntity getTrackedEntity(
+  /**
+   * Gets a tracked entity based on the TE registration org unit
+   *
+   * @return the TE object if found and accessible by the current user
+   * @throws NotFoundException if uid does not exist
+   * @throws ForbiddenException if TE registration org unit not in user's scope or not enough
+   *     sharing access
+   */
+  private TrackedEntity getTrackedEntity(
+      String uid, TrackedEntityParams params, boolean includeDeleted)
+      throws NotFoundException, ForbiddenException {
+    TrackedEntity trackedEntity = getTrackedEntity(uid);
+    UserDetails currentUser = getCurrentUserDetails();
+
     List<String> errors = trackerAccessManager.canRead(currentUser, trackedEntity);
+    if (!errors.isEmpty()) {
+      throw new ForbiddenException(errors.toString());
+    }
+
+    return mapTrackedEntity(trackedEntity, params, currentUser, includeDeleted);
+  }
 
+  /**
+   * Gets a tracked entity based on the program and org unit ownership
+   *
+   * @return the TE object if found and accessible by the current user
+   * @throws NotFoundException if uid does not exist
+   * @throws ForbiddenException if TE owner is not in user's scope or not enough sharing access
+   */
+  private TrackedEntity getTrackedEntity(
+      String uid, Program program, TrackedEntityParams params, boolean includeDeleted)
+      throws NotFoundException, ForbiddenException {
+    TrackedEntity trackedEntity = trackedEntityStore.getByUid(uid);
+    addTrackedEntityAudit(trackedEntity, CurrentUserUtil.getCurrentUsername());
+    if (trackedEntity == null) {
+      throw new NotFoundException(TrackedEntity.class, uid);
+    }
+
+    UserDetails userDetails = getCurrentUserDetails();
+    List<String> errors =
+        trackerAccessManager.canReadProgramAndTrackedEntityType(
+            userDetails, trackedEntity, program);
     if (!errors.isEmpty()) {
       throw new ForbiddenException(errors.toString());
     }
 
+    String error =
+        trackerAccessManager.canAccessProgramOwner(userDetails, trackedEntity, program, false);
+    if (error != null) {
+      throw new ForbiddenException(error);
+    }
+
+    return mapTrackedEntity(trackedEntity, params, userDetails, includeDeleted);
+  }
+
+  /**
+   * Gets the requested tracked entity if the user has access to any program where it's enrolled in,
+   * or to its registering org unit, in case it's not enrolled in any.
+   *
+   * @return the TE object if found and accessible by the user
+   * @throws NotFoundException if TE does not exist
+   * @throws ForbiddenException if TE is not accessible
+   */
+  private TrackedEntity getTrackedEntity(String uid) throws NotFoundException, ForbiddenException {
+    TrackedEntity trackedEntity = trackedEntityStore.getByUid(uid);
+    addTrackedEntityAudit(trackedEntity, CurrentUserUtil.getCurrentUsername());
+    if (trackedEntity == null) {
+      throw new NotFoundException(TrackedEntity.class, uid);
+    }
+
+    if (programService.getAllPrograms().stream()
```

Comment:
@enricocolasante 
`getProgramsByTrackedEntityType` does not validate metadata access, so I think it's better to leave it as it is now and filter by the TET later on. Do you agree?